File: ejercicio.py
from enum import Enum


class Raza(Enum):
    SUPERGUERRERO: str = "superguerrero"
    HUMANO: str = "humano"
    ANDROIDE: str = "androide"
    NAMEKIANO: str = "namekiano"


class GuerreroZ:

    guerreros = []

    def __init__(self, nombre: str, edad: int, ataques: dict, raza: Raza) -> None:
        self.nombre = nombre
        self.edad = edad
        self.ataques = ataques
        self.raza = raza

        self.guerreros.append(self)

    def ataqueMasPoderosoDeCadaGuerrero(self):
        return max(f"{nombre}:{valor}" for nombre, valor in self.ataques.items())

    @classmethod
    def cuantosPersonajesHayDeCadaRaza(self) -> dict:
        conteoPersonajesPorRaza = {}
        for guerrero in self.guerreros:
            razaActual = guerrero.raza.name
            if razaActual in conteoPersonajesPorRaza:
                conteoPersonajesPorRaza[razaActual] += 1
            else:
                conteoPersonajesPorRaza[razaActual] = 1

        return conteoPersonajesPorRaza

    def __str__(self) -> str:
        return f"Nombre: {self.nombre}\nEdad: {self.edad}\nAtaques: {self.ataques}\nRaza: {self.raza.name}"


goku = GuerreroZ(
    "Goku", 17, {"KameHame": 100, "SuperKameHame": 500}, Raza.SUPERGUERRERO
)

vegeta = GuerreroZ(
    "Vegeta", 17, {"CanoGarlick": 100, "FinalFlash": 500}, Raza.SUPERGUERRERO
)

corpetit = GuerreroZ(
    "Corpetit", 17, {"CanoGarlick": 100, "FinalFlash": 500}, Raza.NAMEKIANO
)

fullet_tortuga = GuerreroZ(
    "Fullet Tortuga", 17, {"CanoGarlick": 100, "FinalFlash": 500}, Raza.HUMANO
)


print(goku)

print(goku.ataqueMasPoderosoDeCadaGuerrero())

print(vegeta.ataqueMasPoderosoDeCadaGuerrero())

print(GuerreroZ.cuantosPersonajesHayDeCadaRaza())

File: ejercicio.php
<?php

enum Raza: string
{
  case HUMANO = "humano";
  case SUPERGUERRERO = "superguerrero";
  case NAMEKIANO = "namekiano";
  case ANDROIDE = "androide";
}

class GuerreroZ
{
  private static $guerreros = [];

  private string $nombre;
  private Raza $raza;
  private string $tipo;
  private array $ataques;

  public function __construct(string $nombre, Raza $raza, string $tipo, array $ataques)
  {
    $this->nombre = $nombre;
    $this->raza = $raza;
    $this->tipo = $tipo;
    $this->ataques = $ataques;

    self::$guerreros[] = $this;
  }

  public function ataqueMasPoderoso(): array
  {
    $ataqueMasPoderoso = null;
    $poderMasAlto = 0;

    foreach ($this->ataques as $nombreAtaque => $poderAtaque) {
      if ($poderAtaque > $poderMasAlto) {
        $ataqueMasPoderoso = $nombreAtaque;
        $poderMasAlto = $poderAtaque;
      }
    }

    if ($ataqueMasPoderoso !== null) {
      return [$ataqueMasPoderoso, $poderMasAlto];
    } else {
      return []; // No hay ataques
    }
  }

  public static function cuantosPersonajesPorRaza(): array
  {
    $conteoPersonajes = [];
    foreach (self::$guerreros as $guerrero) {
      $razaActual = $guerrero->raza->value;
      if ($conteoPersonajes[$razaActual]) {
        $conteoPersonajes[$razaActual]++;
      } else {
        $conteoPersonajes[$razaActual] = 1;
      }
    }
    return $conteoPersonajes;
  }

  public function __toString(): string
  {
    return "Nombre: $this->nombre\nRaza: $this->raza\nTipo: $this->tipo\nAtaques: " . print_r($this->ataques, true);
  }
}


$goku = new GuerreroZ("Goku", Raza::HUMANO, "Superpoder", ["Superpoder 1", "Superpoder 2", "Superpoder 3"]);
echo $goku;
